lib/datasets.go

```diff
@@ -1736,22 +1179,534 @@ type datasetImpl struct{}
+			if err != nil {
+				return nil, err
+			}
+			err = c.Assign(built)
+			if err != nil {
+				log.Error(err)
+			}
+		}
+		refs, err = c.ListRefs()
+		if err != nil {
+			return nil, err
+		}
+		// Filter references so that only with a matching name are returned
+		if p.Term != "" {
+			matched := make([]reporef.DatasetRef, len(refs))
+			count := 0
+			for _, ref := range refs {
+				if strings.Contains(ref.AliasString(), p.Term) {
+					matched[count] = ref
+					count++
+				}
+			}
+			refs = matched[:count]
+		}
+		// Filter references by skipping to the correct offset
+		if p.Offset > len(refs) {
+			refs = []reporef.DatasetRef{}
+		} else {
+			refs = refs[p.Offset:]
+		}
+		// Filter references by limiting how many are returned
+		if p.Limit < len(refs) {
+			refs = refs[:p.Limit]
+		}
+		// TODO(dlong): Filtered by p.Published flag
+	} else if listProfile.Peername == "" || reqProfile.Peername == listProfile.Peername {
+		refs, err = base.ListDatasets(scope.Context(), scope.Repo(), p.Term, p.Offset, p.Limit, p.RPC, p.Public, p.ShowNumVersions)
+		if errors.Is(err, ErrListWarning) {
+			listWarning = err
+			err = nil
+		}
+	} else {
+		return nil, fmt.Errorf("listing datasets on a peer is not implemented")
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	if p.EnsureFSIExists {
+		// For each reference with a linked fsi working directory, check that the folder exists
+		// and has a .qri-ref file. If it's missing, remove the link from the centralized repo.
+		// Doing this every list operation is a bit inefficient, so the behavior is opt-in.
+		for _, ref := range refs {
+			if ref.FSIPath != "" && !linkfile.ExistsInDir(ref.FSIPath) {
+				ref.FSIPath = ""
+				if ref.Path == "" {
+					if err = scope.Repo().DeleteRef(ref); err != nil {
+						log.Debugf("cannot delete ref for %q, err: %s", ref, err)
+					}
+					continue
+				}
+				if err = scope.Repo().PutRef(ref); err != nil {
+					log.Debugf("cannot put ref for %q, err: %s", ref, err)
+				}
+			}
+		}
+	}
+	// Convert old style DatasetRef list to VersionInfo list.
+	// TODO(dlong): Remove this and convert lower-level functions to return []VersionInfo.
+	infos := make([]dsref.VersionInfo, len(refs))
+	for i, r := range refs {
+		infos[i] = reporef.ConvertToVersionInfo(&r)
+	}
+
+	if listWarning != nil {
+		return nil, listWarning
+	}
+
+	return infos, nil
+}
+
+// ListRawRefs gets the list of raw references as string
+func (datasetImpl) ListRawRefs(scope scope, p *ListParams) (string, error) {
+	text := ""
+	if p.UseDscache {
+		c := scope.Dscache()
+		if c == nil || c.IsEmpty() {
+			return "", fmt.Errorf("repo: dscache not found")
+		}
+		text = c.VerboseString(true)
+		return text, nil
+	}
+	return base.RawDatasetRefs(scope.Context(), scope.Repo())
+}
+
+// Get retrieves datasets and components for a given reference.t
+func (datasetImpl) Get(scope scope, p *GetParams) (*GetResult, error) {
+	if err := qfs.AbsPath(&p.Outfile); err != nil {
```

Comment:
Ah, another caveat of `Dispatch`: this call has to happen in the method, not the implementation. The reason is that, when `qri connect` is running, the caller may pass a relative path which needs to be turned into an absolute path before being sent over RPC. Otherwise, the caller and receiver may be in running in different directories and the relative path won't work.

See `Status` in `lib/fsi.go` for an example of how this is done.

I'm of the opinion that `Dispatch` should be made aware of which parameter fields are filepaths, and it should be calling `AbsPath` as needed. But the code to do this doesn't exist yet.